tests/structured_conf/test_structured_config.py

```diff
@@ -1070,3 +1070,176 @@ def test_merge_missing_list_promotes_target_type(self, class_type: str) -> None:
+                },
+            ),
+            (
+                "ComplexList",
+                {"list": [[{"foo": 13}]]},
+                {"list": [[{"foo": 13}]]},
+            ),
+            (
+                "ComplexDict",
+                {"dict": {"foo2": {"var2": [3, 4]}}},
+                {"dict": {"foo": {"var": [1, 2]}, "foo2": {"var2": [3, 4]}}},
+            ),
+        ],
+    )
+    def test_container_as_element_type_merge(
+        self, class_type: str, obj: str, obj2: Any, expected: Any
+    ) -> None:
+        module: Any = import_module(class_type)
+        class_ = getattr(module, obj)
+        cfg = OmegaConf.merge(class_)
+        res = OmegaConf.merge(cfg, obj2)
+        assert res == expected
+
+    @pytest.mark.parametrize(  # type: ignore
+        "obj,obj2",
+        [
+            (
+                "ContainerInList",
+                {
+                    "list_with_list": [[4, "invalid_value"]],
+                    "list_with_dict": [{"foo": 4}],
+                },
+            ),
+            (
+                "ContainerInList",
+                {"list_with_list": [[4, User()]], "list_with_dict": [{"foo": 4}]},
+            ),
+            (
+                "ContainerInList",
+                {
+                    "list_with_list": [[4, 5]],
+                    "list_with_dict": [{"foo": "invalid_value"}],
+                },
+            ),
+            (
+                "ContainerInList",
+                {"list_with_list": [[4, 5]], "list_with_dict": [{"foo": User()}]},
+            ),
+            (
+                "ContainerInDict",
+                {
+                    "dict_with_list": {"foo": [3, "invalid_value"]},
+                    "dict_with_dict": {"foo": {"var": 4}},
+                },
+            ),
+            (
+                "ContainerInDict",
+                {
+                    "dict_with_list": {"foo": [3, User()]},
+                    "dict_with_dict": {"foo": {"var": 4}},
+                },
+            ),
+            (
+                "ContainerInDict",
+                {
+                    "dict_with_list": {"foo": [3, 4]},
+                    "dict_with_dict": {"foo": {"var": "invalid_value"}},
+                },
+            ),
+            (
+                "ContainerInDict",
+                {
+                    "dict_with_list": {"foo": [3, 4]},
+                    "dict_with_dict": {"foo": {"var": User()}},
+                },
+            ),
+            (
+                "ComplexList",
+                {
+                    "list": [[{"foo": "invalid"}]],
+                },
+            ),
+            (
+                "ComplexList",
+                {
+                    "list": [[{"foo": User()}]],
+                },
+            ),
+            (
+                "ComplexDict",
+                {
+                    "dict": {"foo2": {"var2": ["invalid", 4]}},
+                },
+            ),
+        ],
```

Comment:
try to remove trialing comma from lines like:
```
 "list": [[{"foo": "invalid"}]],
```

I will likely result in a more compact formatting.